Use Snyk Container Monitor in the Snyk GitHub action

lmath opened the pull request:
We are moving over to using `snyk container monitor` because we want to
- be able to keep track of vulnerabilities in the docker images we create on release without manually adding them in the Snyk UI
- continue scanning our java/scala jar files

So this change updates the Snyk github action to:
- create local docker images
- scan those local images with `snyk container monitor`

when a tag is pushed that is not an rc tag.

lmath commented:
#### Re: does the Snyk action scan the latest application dependencies?

We were previously manually updating the latest version of docker images that had been released to ensure that application dependencies would be re-imported for the scan. So, how can we be convinced that with this change that we are still scanning the latest application dependencies and the docker image that will be published in the release?

We will compare the output to the Snyk dashboard when this action runs for this branch:
![image](https://user-images.githubusercontent.com/3072877/218501630-709f4c4e-9489-4a99-a148-602ab2575af5.png)

With the output after we temporarily introduce a dependency that we expect will increase vulnerabilities on the dashboard
![image](https://user-images.githubusercontent.com/3072877/218501819-8e9249dd-1012-4959-bb35-2628c0e65b0f.png)

And indeed, when we look on the Snyk dashboard, we see it appear. 
![image](https://user-images.githubusercontent.com/3072877/218504238-461d6268-c7e3-420e-bb3f-17541b7dbfb5.png)

lmath commented:
### If this updated Snyk action runs more than once, will there be duplicates in the dashboard?

We require the dashboard to just show the latest versions. 

At 11:36 today I pushed tag `5.3.2-rc9`
![image](https://user-images.githubusercontent.com/3072877/218728669-fb1de4af-b600-4cb0-bd81-816d1520a5ab.png)

and we saw 5 projects appear for RDB loader
![image](https://user-images.githubusercontent.com/3072877/218729251-a7cc17dd-dcea-4e2e-9832-94a3be623516.png)

and we can see if we click through on one of the docker images that it has scanned tag `5.3.2-rc9`
![image](https://user-images.githubusercontent.com/3072877/218729451-72c06c00-b6e3-4650-86be-ca1a7905fcb2.png)

When I push tag `5.3.2-rc10` at 11:48
![image](https://user-images.githubusercontent.com/3072877/218730944-6602ccc3-a5af-4fb6-8d5a-542aa08a7956.png)

And the Snyk github action completes, I can see that in the new dashboard we still just have the 5 projects for RDB loader as expected
![image](https://user-images.githubusercontent.com/3072877/218731220-357f5d05-29ee-471c-b975-39bc1e6105d9.png)

and that e.g. the project for `rdb-loader-databricks` has re run for tag `5.3.2-rc10` 
![image](https://user-images.githubusercontent.com/3072877/218731476-c961e9f0-9c53-401f-87f2-d9afec8d6c96.png)

lmath commented:
Note: around 11:30am UK time on 14 Feb we updated the Snyk token for RDB loader so that Snyk reports will show up in the new prod org.

I pushed 5.3.2-rc11 to double check that the Snyk action still worked after I removed the step in `snyl.yml` that was running `docker image ls` as it was just there while I was debugging, and after I tidied up the git commits.

I then amended the commit once more to ensure that [this](https://github.com/snowplow/snowplow-rdb-loader/pull/1187/files#diff-279a5a5997dd6e7a67267df6229cd45adb7acdb39c8ebab573f98e524deb0fd5R10) if condition checks that the tag does not have "rc" in it, and then pushed tag `5.3.2-rc12`. 

I then validated that the Snyk action did not run for `5.3.2-rc12` as I'd expect after the change to the if condition:
<img width="1502" alt="image" src="https://user-images.githubusercontent.com/3072877/218785360-1f90d61b-fa34-4126-9518-f3df6e477339.png">

istreeter commented:
Thanks @lmath!  Just a couple of questions....

Did you consider whether this should be its own github action vs adding it as a new step of the ci action?  Currently, this new action needs to compile all the code and build the jar files -- so it is repeating work that is already being done by the ci action.  I guess that's not a huge problem, but I wanted to mention it.

And just for curiosity... what would happen if someone were to run the snyk command locally, outside of the github action?  E.g. if I ran `snyk container monitor --app-vulns snowplow/snowplow-transformer-kinesis:5.3.1` using my own snyk token.  Would this update the project in snyk, or would it create a new project.  It might be helpful for us to know how to manually run a scan so that it updates the project.

spenes reviewed (approved):
Thanks @lmath! I agree with Ian's comment about making this part of ci action to avoid building project one more time. Aside from this, I think it looks good 👍

lmath commented:
### Why are we creating local docker images in a snyk.yml instead of scanning the images created in ci.yml?

Ideally, since we are already building images in `ci.yml` it does make sense to just scan those! I thought so too and tried to do that initially, and later found out that it didn't work. Here's why. We can see in the [README](https://github.com/docker/build-push-action) for the docker push action that it is possible to set the `load` parameter to true or false, where setting it to true is shorthand for `--output=type=docker`. Setting it to true loads the image into docker locally so that you can see the image if you run `docker image ls`. This sounds like exactly what we want! Well, that plus continuing to push the images to the remote docker registry...

Alas, loading images into docker locally isn't currently supported for multi-platform builds like we have here in RDB loader. 

https://docs.docker.com/engine/reference/commandline/buildx_build/#load

![image](https://user-images.githubusercontent.com/3072877/219010547-b37a7d7f-826d-4121-9f53-c9dcdf333ff6.png)

You can see we are not the only ones interested in this support by checking out [this github issue](https://github.com/docker/buildx/issues/59).

Since loading to local docker isn't supported for multiplatform images, for the moment, we'd have the build the images twice either way. And while that remains the case, I thought it was tidier to keep the Snyk Github action separate. 

#### What if we just tried it? What kind of error do we get? 🤔 
I lost track of the run where I tried this initially, so I redid it to note down what happens here.

I set both push and load to true to try to ensure the image gets loaded locally AND gets pushed to the remote docker registry like this: 
![image](https://user-images.githubusercontent.com/3072877/219013216-74192dd8-5e1a-47f0-8a35-f8e6e5f4fc9c.png)

then I pushed a new tag `5.3.2-rc13` and you can see the github action run [here](https://github.com/snowplow/snowplow-rdb-loader/actions/runs/4183157703/jobs/7247360844)

Here is an example error and `docker buildx` command that was generated:
<img width="1034" alt="image" src="https://user-images.githubusercontent.com/3072877/219013611-4db4f2ca-8d74-47fe-9d73-5ef6348f96e4.png">

thanks for reminding me to explain this @istreeter @spenes

lmath commented:
@istreeter 

I just tried out running first 
`snyk container monitor --app-vulns snowplow/transformer-kinesis:5.3.1-distroless` 

which then updated the existing transformer kinesis project (for the distroless image - I first ran it for the image that isn't distroless and saw a duplicate until I realised why).

And then I ran
`snyk container monitor --app-vulns snowplow/transformer-kinesis:5.3.2-rc11-distroless`
which scanned `snowplow/transformer-kinesis:5.3.2-rc11-distroless` before updating the same transformer kinesis project

I am using the same API token I was using previously and running the CLI locally caused the prod org to update, but I was relying on my preferred org in my settings being the prod org. Probably it's best to pass the org id parameter, both if running `snyk container monitor ...` locally and via the Github action. 

I just updated the commit on this branch to pass the org param to the github action (and disabled the if condition for testing), then pushed a new tag `5.3.2-rc14`. Indeed, it has run successfully (I checked the [corresponding Github action run](https://github.com/snowplow/snowplow-rdb-loader/actions/runs/4185226946) and checked the prod Snyk UI to see it had been updated). So I have updated the commit to include the org param.

istreeter commented:
Hi @lmath -- I understand the problems with trying to combine snyk with the images built in the existing ci action.  But I was trying to suggest something different.

Basically... in the ci action, you could run `sbt docker:publishLocal` _in addition to_ the existing docker build we already do.  Meaning, append it [after this line](https://github.com/snowplow/snowplow-rdb-loader/blob/5.3.1/.github/workflows/ci.yml#L156).  Then, you will have a well-behaved local docker image for which you can safely run the snyk command.

lmath commented:
### Comparing running times of running the Snyk action separately vs adding it to `ci.yml`
Since we needed to create images twice either way, I did not understand the benefits of expanding `ci.yml` instead of having a separate `snyk.yml` as @istreeter suggested. So I decided to just try it out and and compare the running times 🧑‍🔬 

When I pushed `5.3.2-rc14`, this created local docker images and scanned then with Snyk in a separate action `snyk.yml`.

![image](https://user-images.githubusercontent.com/3072877/219646045-e177b59f-6329-4a6f-b633-dce0c46835d4.png)

`5.3.2-rc11` which also created local docker images and scanned then with Snyk in a separate action `snyk.yml`, but didn't pass the `--org` param took a similar amount of time.
![image](https://user-images.githubusercontent.com/3072877/219687933-6ab3d340-a15e-45fa-b0c6-86f06dbb8a9c.png)


When I instead appended those steps to create local docker images and then scan them with Snyk into the existing `ci.yml`
![image](https://user-images.githubusercontent.com/3072877/219681076-b48f0dde-c896-41a7-834a-2154d0c88278.png)

and it appears that a `ci.yml` that adds in creating local docker images and scanning them doesn't seem to increase the overall running time of `ci.yml`. I re-ran the code with an extended `ci.yml` with a snyk scan just to make sure that the `5.3.2-rc16` wasn't an anomaly. 
![image](https://user-images.githubusercontent.com/3072877/219688205-fe213d03-0c28-4cb0-91e0-1c82ef3f08b4.png)

I thought having `snyk.yml` decoupled was nice, but it's not nice enough to add 3-4 minutes of running time, so I'll move the `docker:publishLocal` and snyk scan steps over to `ci.yml`

Note: 
- I did check the Snyk projects dashboard to see that each RDB loader projects had updated as expected around the time the Github action had finished running.
- After I updated the if checks to only generate local docker images and scan them with snyk if the tag is not an rc tag,  I pushed `5.3.2-rc18` and checked those two steps would be skipped
![image](https://user-images.githubusercontent.com/3072877/219696207-7bff17e8-a55b-465c-bd18-3f6f965071d7.png)

istreeter reviewed (approved):
This looks great, let's merge it in.
